src/app/fdctl/run/tiles/fd_nettx.c

```diff
@@ -0,0 +1,302 @@
+  if( FD_LIKELY( (handled_frag || !ctx->frame_cnt) && ctx->frame_cnt<1024UL ) ) return;
+
+  *charge_busy = 1;
+
+  int sent_cnt = sendmmsg( ctx->send_fd, ctx->hdr+ctx->frame_start, (uint)ctx->frame_cnt, MSG_DONTWAIT );
+  if( FD_UNLIKELY( -1==sent_cnt && errno==EAGAIN ) ) {
+    /* Internal frame buffer is full and couldn't send anything, we are
+       now backpressured by the socket and shouldn't poll any more. */
+    *opt_poll_in = 2;
+  } else if( FD_UNLIKELY( -1==sent_cnt ) ) {
+    FD_LOG_ERR(( "sendmmsg failed (%d-%s)", errno, fd_io_strerror( errno ) ));
+  }
+  FD_TEST( sent_cnt ); /* Should not be zero, -1 indicates error */
+
+  ctx->frame_start = (ctx->frame_start+(ulong)sent_cnt)%1024UL;
+  ctx->frame_cnt -= (ulong)sent_cnt;
+}
+
+static inline int
+before_frag( fd_nettx_ctx_t * ctx,
+             ulong            in_idx,
+             ulong            seq,
+             ulong            sig ) {
+  (void)in_idx;
+  (void)seq;
+
+  ctx->handled_frag = 1;
+  return fd_disco_netmux_sig_dst_ip( sig )%ctx->round_robin_cnt != ctx->round_robin_id;
+}
+
+static inline void
+during_frag( fd_nettx_ctx_t * ctx,
+             ulong            in_idx,
+             ulong            seq,
+             ulong            sig,
+             ulong            chunk,
+             ulong            sz ) {
+  (void)in_idx;
+  (void)seq;
+  (void)sig;
+
+  if( FD_UNLIKELY( chunk<ctx->in[ in_idx ].chunk0 || chunk>ctx->in[ in_idx ].wmark || sz>FD_NET_MTU ) )
+    FD_LOG_ERR(( "chunk %lu %lu corrupt, not in range [%lu,%lu]", chunk, sz, ctx->in[ in_idx ].chunk0, ctx->in[ in_idx ].wmark ));
+
+  uchar * src = (uchar *)fd_chunk_to_laddr( ctx->in[ in_idx ].mem, chunk );
+
+  ulong frame_idx = (ctx->frame_start+ctx->frame_cnt)%1024UL;
+  fd_memcpy( ctx->frame[ frame_idx ], src, sz );
+}
+
+static inline void
+after_frag( fd_nettx_ctx_t *    ctx,
+            ulong               in_idx,
+            ulong               seq,
+            ulong               sig,
+            ulong               chunk,
+            ulong               sz,
+            ulong               tsorig,
+            fd_stem_context_t * stem ) {
+  (void)in_idx;
+  (void)seq;
+  (void)sig;
+  (void)chunk;
+  (void)tsorig;
+  (void)stem;
+
+  ulong frame_idx = (ctx->frame_start+ctx->frame_cnt)%1024UL;
+  struct mmsghdr * hdr         = ctx->hdr+frame_idx;
+  struct sockaddr_in * saddr4  = fd_type_pun( hdr->msg_hdr.msg_name );
+
+  uchar const *        frame    = ctx->frame[ frame_idx ];
+  uchar const *        cur      = frame;
+  fd_eth_hdr_t const * eth_hdr  = fd_type_pun_const( cur );  cur += sizeof(fd_eth_hdr_t);
+  fd_ip4_hdr_t const * ip4_hdr  = fd_type_pun_const( cur );  cur += FD_IP4_GET_IHL( *ip4_hdr )<<2;
+  memcpy( &saddr4->sin_addr.s_addr, &ip4_hdr->daddr_c, 4UL );
+
+  uchar const *         data    = cur;
+  long                  data_sz = (long)sz - (long)((ulong)cur - (ulong)frame);
+  if( FD_UNLIKELY( eth_hdr->net_type!=fd_ushort_bswap( FD_ETH_HDR_TYPE_IP ) ||
```

Comment:
isn't this an uninitialied read? looks like a read is done here before data_sz is checked.
I suggest splitting this up into separate branches